felix/bpf/nat/connecttime.go

```diff
@@ -125,21 +130,47 @@ func loadProgram(logLevel, ipver string, udpNotSeen time.Duration, excludeUDP bo
 	return obj, nil
 }
 
-func attachProgram(name, ipver, bpfMount, cgroupPath string, udpNotSeen time.Duration, excludeUDP bool, obj *libbpf.Obj) error {
-
-	progPinDir := path.Join(bpfMount, "calico_connect4")
-	_ = os.RemoveAll(progPinDir)
-
+func attachProgram(name, ipver, bpfMount, cgroupPath string, udpNotSeen time.Duration, excludeUDP bool, obj *libbpf.Obj, legacy bool) error {
 	progName := "calico_" + name + "_v" + ipver
+	progPinPath := path.Join(bpfMount, progName)
+	if _, err := os.Stat(progPinPath); err == nil {
+		link, err := libbpf.OpenLink(progPinPath)
+		if err != nil {
+			return fmt.Errorf("error opening link %s : %w", progPinPath, err)
+		}
+		defer link.Close()
+		if err := link.Update(obj, progName); err != nil {
+			return fmt.Errorf("error updating program %s : %w", progName, err)
+		}
+		return nil
+	}
 
-	// N.B. no need to remember the link since we are never going to detach
-	// these programs unless Felix restarts.
-	if _, err := obj.AttachCGroup(cgroupPath, progName); err != nil {
-		return fmt.Errorf("failed to attach program %s: %w", progName, err)
+	// Used only for UT to test legacy way to attach.
+	if legacy {
+		err := obj.AttachCGroupLegacy(cgroupPath, progName)
+		if err != nil {
+			return fmt.Errorf("failed to attach program %s: %w", progName, err)
+		}
+		return nil
 	}
 
+	// Tries non-legacy and fallsback to legacy if non-legacy fails.
+	link, err := obj.AttachCGroup(cgroupPath, progName)
+	if err != nil {
+		err = obj.AttachCGroupLegacy(cgroupPath, progName)
+		if err != nil {
+			return fmt.Errorf("failed to attach program %s: %w", progName, err)
+		}
+		link = nil
+	}
+	if link != nil {
```

Comment:
```suggestion
	} else if link != nil {
```

assigning nil to the link is quite confusing